### ISSUE TITLE
sokol-app emscripten: right mouse button doesn't seem to work

### ISSUE THREAD

**Issue by floooh** [floooh/sokol-samples#8]
See imgui-sample: https://floooh.github.io/sokol-html5/wasm/imgui-emsc.html under "Popups & Modal windows" -> Context menus

**Comment by RobertoMalatesta** [comment]
Mousedown returns  b2 on your web demos instead of b1 (normal imgui).
(see Input, Navigation and Focus -> Keyboard, Mouse and Navigation State)

--R

**Comment by floooh** [comment]
Yep I debugged and fixed the RMB problem today, currently only in the sokol_app.h samples though, and the web page hasn't been updated yet.

Unfortunately I have no clue what would cause the high CPU usage, on my MBP it's also fairly high, but also in other WebGL demos, the clear-sample which basically does nothing on the CPU has the same usage as the ImGui-demos :/

<img width="362" alt="screen shot 2018-05-30 at 19 46 42" src="https://user-images.githubusercontent.com/1699414/40737894-508dab72-6442-11e8-8436-ef52aab87d36.png">

<img width="366" alt="screen shot 2018-05-30 at 19 46 01" src="https://user-images.githubusercontent.com/1699414/40737896-5247bf8e-6442-11e8-9f04-21112f6a3077.png">

**Comment by floooh** [comment]
Ok, I've committed a fix for the right mouse button to the 'normal' emscripten samples to, and updated the Sokol-Samples webpage: https://floooh.github.io/sokol-html5/

The yet 'unofficial' sapp-samples here: http://floooh.github.io/oryol-sticky-tests/imgui-highdpi-sapp.html hasn't been updated, but that's happen either when I do other fixes, or later when sokol-app I will integrate those samples into the 'official' Sokol-Samples webpage :)

**Comment by RobertoMalatesta** [comment]
Hi @floooh ,



on the CPU high usage I suspect a structural refreshing issue.
Thought it could be ImGUI but here's a wasm port with almost 0 cpu usage I found some months ago and deployed in a test dir on my site.
 (don't remember which github fork now, but I will give due credits to the author as soon as I collect the info) 
See:
[https://www.exedraos.com/2018_test/prova02/clicca_qui.html](https://www.exedraos.com/2018_test/prova02/clicca_qui.html)

When you use the pause check the GUI is still responsive and cpu/GPU time goes to almost null.

If it suits you, I'd open a specific issue on this to keep track, as well as some other issues (dragging on tablets/mobiles again not working, etc)


--R

**Comment by floooh** [comment]
Ah ok, I see, that demo seems to have some special pause mode, where rendering only happens on input events (seems to be on mouse click, since mouse hovering doesn't work in pause).  This isn't some sort of built-in ImGui-feature, but instead how the UI application uses ImGui.

My demos basically don't have that special paused-handling.

Outside pause mode, CPU usage in that demo is slightly higher (could be because of the additional 3D rendering):
<img width="356" alt="screen shot 2018-05-31 at 08 44 49" src="https://user-images.githubusercontent.com/1699414/40766422-b3704516-64af-11e8-9acd-327f9f05c52c.png">

This is the wasm demo from the sokol-samples page:
<img width="356" alt="screen shot 2018-05-31 at 08 46 39" src="https://user-images.githubusercontent.com/1699414/40766431-bb824358-64af-11e8-86c1-48e25a390848.png">

**Comment by RobertoMalatesta** [comment]
> an reach nearly 0 CPU usage is by stopping the entire WebGL rendering as well 

If you check "Rendering when mouse up" you 'll notice that rendering is done at mouse movement.
This greatly improves the cpu draining issue.

> I think that's a very specific requirement and only useful for pure UI applications that don't have any animations. 
...
For mobile power usage it would be very useful though, but for every case where ImGui is overlayed over a dynamic 3D scene it would be quite worthless.

Optimizing CPU/GPU usage will leave more power to other threads (yes, I'm waiting to have them in emscripten since 2010) i.e. to run some DL like dLib or Tensorflow tasks.
Besides, if your 3d animated models happens not to change screen position for a while, that's a plain heat waste, especially on mobile unless you want to use your device for handkerchiefs ironing 

--R

PS: pardon my ignorance-driven following question but, since glfw3 is still broken at managing touch inputs, does Sokol works with SDL2 ?

**Comment by floooh** [comment]
I haven't tested with SDL2. For touch input I'm planning to support the low-level touches-begin/changed/end/cancel events in the new sokol_app.h header (these are fairly portable across iOS, Android and Mobile Web), and then have a separate higher-level sokol_input.h header, which would add input polling and gesture recognition on top of the sokol_app.h event system (the feature set will be similar to what I'm doing here in Oryol: http://floooh.github.io/oryol/wasm/TestInput.html).

However, I never got touch input working well with ImGui so far, mostly not recognizing buttons on tapping. I suspect it's the same problem which I only fixed now in the imgui-sapp sample, where ImGui cannot know about up/down input when it happens in the same frame, so some code outside ImGui needs to queue events. I'm going to fix the native iOS code in sokol_app.h next, and once I got ImGui touch input working, do the same for HTML touch input.
